### Imports

In [1]:
import numpy as np
from numpy import array, log, exp, where, vectorize
from bond_pricing import bond_duration, duration
from datetime import datetime

### Standard Inputs

In [2]:
coupon = 0.032
par = 100
ytm = 0.032
settle = datetime(2025, 10, 15)
mat = datetime(2030, 10, 15)
freq = 2

### Macaulay Duration

$$MacDur=\left\lbrace\frac{1+r}{r}-\frac{1+r+\left\lbrack N\times\left(c-r\right)\right\rbrack}{c\times\left\lbrack\left(1+r\right)^{N}-1+r\right\rbrack}\right\rbrace-\frac{t}{T}$$

- *r* is the yield-to-maturity per period;
- *N* is the number of evenly spaced periods to maturity as of the beginning of the current period;
- *c* is the coupon rate per period;
- *t* is the number of days from the last coupon payment to the settlement date; and
- *T* is the number of days in the coupon period.

In [12]:
print(bond_duration(settle=settle, cpn=coupon, mat=mat, yld=ytm, freq=freq, face=par))

4.660146890283194


In [15]:
n = (mat.year - settle.year)*2
arr = np.full(n, coupon/2 * par)
arr[n-1] += par
print(list(arr))

duration(cf=arr, rate=ytm, cf_freq=2, comp_freq=2)

[1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 101.6]


4.660146890283189

| Period | Time to Reciept | Cash Flow | PV       | Weight       | Time x Weight |
| ------ | --------------- | --------- | -------- | ------------ | ------------- |
| 1      | 1               | 1.6       | 1.5748   | 0.0157       | 0.0157        |
| 2      | 2               | 1.6       | 1.5500   | 0.0155       | 0.0310        |
| 3      | 3               | 1.6       | 1.5256   | 0.0153       | 0.0458        |
| 4      | 4               | 1.6       | 1.5016   | 0.0150       | 0.0601        |
| 5      | 5               | 1.6       | 1.4779   | 0.0148       | 0.0739        |
| 6      | 6               | 1.6       | 1.4546   | 0.0145       | 0.0873        |
| 7      | 7               | 1.6       | 1.4317   | 0.0143       | 0.1002        |
| 8      | 8               | 1.6       | 1.4092   | 0.0141       | 0.1127        |
| 9      | 9               | 1.6       | 1.3870   | 0.0139       | 0.1248        |
| 10     | 10              | 101.6     | 86.6875  | 0.8669       | 8.6688        |
|        |                 |           | 100.0000 | 1.0000       | 9.3203        |
|        |                 |           |          | Mac Duration | 4.6601        |
|        |                 |           |          | Mod Duration | 4.5868        |

In [18]:
n = (mat.year - settle.year)*2

cf = np.full(n, coupon/2 * par)
cf[n-1] += par
print(list(cf))



#cf_t = array()

[1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 101.6]


### Modified Duration

In [16]:
print(bond_duration(settle=settle, cpn=coupon, mat=mat, yld=ytm, freq=freq, face=par, modified=True))

4.586758750278734


In [17]:
n = (mat.year - settle.year)*2
arr = np.full(n, coupon/2 * par)
arr[n-1] += par
print(list(arr))

duration(cf=arr, rate=ytm, cf_freq=2, comp_freq=2,modified=True)

[1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 1.6, 101.6]


4.586758750278729